### ISSUE TITLE
"Error: Failed to replace env in config: ${NPM_API_KEY}" in after_deploy

### ISSUE THREAD

**Issue by adamjmcgrath** [travis-ci/dpl#298]
.travis.yml

```yaml
language: node_js

node_js:
  - '0.12'

deploy:
  provider: npm
  email: adamjmcgrath@gmail.com
  api_key:
    secure: p2RtKFppPgLZME2ceK8SROmpXcreGoBeqpxJIee3Y5a2XXOa+ughWAN/tcvO8gTngVHYr2XEU+TlDpidC1KGRTZv/4cx1WH0XwnDw1gpcNB3v8G6znx0X0i9XgBCH/HlzLbW2+/zIVVqfzBW3Z2YMf9nen9aO/LhvlOeiu4B4FO2o54ODYcpQO6jdZLhSf7aC+ua29ixvTOG97Z31DUxRI+Idg83VxbCWNEfKmZuxDPrabZhKIDUZK/3LqsmF0HE11kAe/a4lHD0mjO+DAQpWXkauP6FeeKKDnDASs7Sc/XdWCEFTuP9JtN1d7XWTRKPBj/1D62+LCp6jHwzSFPxM6oO7i63iP6y7Fqgod9Okh6qvwkn3nKmXm7WmQJbLS7VGTNQ/G5O6HlQVgmnYRBh3iRptYUHF9wmBcdfXTVQ4bdL7QLVUuzaIBlOGv/9O5M0a19erkq+xQUrasZpnVRW+BcIjbNwNoLRFitDSK6BlRrItj2TT2OOokA9XqMorjMJl9IXReGmKDD1cGctPP1KfSOUEjnollP5JX7IgNf1O13ZRqjfkTuWvCgmXsGN52iD8+CP1KD9SMeu9vH99tMxXIl7JLgjkzZ8o4Z+LZda0Qf3kvCIYikz17G+5hDOalaaDktvDmjEzshFL0riR+SlbVWf9rs1Z3+bXVRmK7VM0go=
  on:
    tags: true
    repo: adamjmcgrath/react-native-simple-auth

after_deploy:
  - openssl aes-256-cbc -K $encrypted_ed11b53bbeb9_key -iv $encrypted_ed11b53bbeb9_iv -in .travis/deploy_key.enc -out .travis/deploy_key -d
  - npm version --no-git-tag-version minor
  - git commit -am "Bump minor for next release [ci skip]"
  - chmod 600 .travis/deploy_key.pem
  - ssh-add .travis/deploy_key
  - git remote add deploy git@github.com:adamjmcgrath/react-native-simple-auth.git
  - git push deploy
```
https://github.com/adamjmcgrath/react-native-simple-auth/blob/master/.travis.yml

Error:

```
$ npm version --no-git-tag-version minor
Error: Failed to replace env in config: ${NPM_API_KEY}
    at /home/travis/.nvm/versions/node/v0.12.7/lib/node_modules/npm/lib/config/core.js:429:13
    at String.replace (native)
    at envReplace (/home/travis/.nvm/versions/node/v0.12.7/lib/node_modules/npm/lib/config/core.js:424:12)
    at parseField (/home/travis/.nvm/versions/node/v0.12.7/lib/node_modules/npm/lib/config/core.js:400:7)
    at /home/travis/.nvm/versions/node/v0.12.7/lib/node_modules/npm/lib/config/core.js:338:17
    at Array.forEach (native)
    at Conf.add (/home/travis/.nvm/versions/node/v0.12.7/lib/node_modules/npm/lib/config/core.js:337:23)
    at ConfigChain.addString (/home/travis/.nvm/versions/node/v0.12.7/lib/node_modules/npm/node_modules/config-chain/index.js:244:8)
    at Conf.<anonymous> (/home/travis/.nvm/versions/node/v0.12.7/lib/node_modules/npm/lib/config/core.js:325:10)
    at fs.js:334:14
```
https://travis-ci.org/adamjmcgrath/react-native-simple-auth/builds/74321653

**Comment by BanzaiMan** [comment]
The log doesn't show the API key decrypted, so I suspect your key was incorrectly encrypted. Retry with `travis encrypt --org -r adamjmcgrath/react-native-simple-auth…`

**Comment by adamjmcgrath** [comment]
The `NPM_API_KEY` seems to have been encrypted correctly because the deploy step managed to publish the library:

```
Deploying application
...
+ react-native-simple-auth@0.2.1
```

It's just that using an `npm` command in the `after_deploy` step isn't working. I suspect that the npm API key has been cleaned up after the `deploy` step and is causing an additional `npm` command in the `after_deploy` step to fail

**Comment by BanzaiMan** [comment]
How are you expecting `$NPM_API_KEY` value to be set? When dpl deploys to NPM, it spawns a new process that sets the environment variable (https://github.com/travis-ci/dpl/blob/70748f8a0ec35838e9c2d3e25f540f9ff16b21b4/lib/dpl/provider/npm.rb#L27). This variable is not available for the build script itself (and the commands running in `after_script`), so if you want to use the value, you will need to set it up yourself.

**Comment by adamjmcgrath** [comment]
> How are you expecting $NPM_API_KEY value to be set?

I expect the `NPM_API_KEY` to be set by the `api_key` field of the deploy yaml and, as you described, only to be available to the `npm publish` command that you linked to:
https://github.com/travis-ci/dpl/blob/70748f8a0ec35838e9c2d3e25f540f9ff16b21b4/lib/dpl/provider/npm.rb#L27

> This variable is not available for the build script itself (and the commands running in after_script), so if you want to use the value, you will need to set it up yourself.

I realise that - I don't want to use the `NPM_API_KEY` env variable in the `after_script` command. I just want to run an npm command (`npm version` doesn't require an api key). The problem is that the deploy step created an .npmrc that will cause subsequent npm commands to fail, unless the user sets the `NPM_API_KEY` env variable.

**Comment by BanzaiMan** [comment]
Very interesting, so `npm` is showing a completely unrelated error message.

**Comment by BanzaiMan** [comment]
I've released 1.7.19.

**Comment by adamjmcgrath** [comment]
Looks good, thanks

The deploy worked:
```
Deploying application
NPM API key format changed recently. If your deployment fails, check your API key in ~/.npmrc.
http://docs.travis-ci.com/user/deployment/npm/
+ react-native-simple-auth@0.2.3
```

And the npm version worked in the after_deploy step:
```
after_deploy.2
0.44s$ npm version --no-git-tag-version minor
v0.3.0
```

https://travis-ci.org/adamjmcgrath/react-native-simple-auth/builds/76582139
